File: day23_mediator/guru/cpp/mediator.cpp
#include "mediator.hpp"

#include <iostream>
#include <string>

#include "component.hpp"


void ConcreteMediator::Notify(const BaseComponent& sender, const std::string& event) const {
  if (event == "A") {
    std::cout << "Mediator reacts on A and triggers following operations:" << '\n';
    component_2_.DoC();
  } else if (event == "D") {
    std::cout << "Mediator reacts on D and triggers following operations:" << '\n';
    component_1_.DoB();
    component_2_.DoC();
  }
}

ConcreteMediator::ConcreteMediator(Component1& component_1, Component2& component_2)
    : component_1_(component_1), component_2_(component_2) {
  component_1_.SetMediator(this);
  component_2_.SetMediator(this);
}

File: day23_mediator/guru/py/mediator.py
from __future__ import annotations
from abc import ABC, abstractmethod
from overrides import overrides
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from component import BaseComponent, Component1, Component2


class Mediator(ABC):
    """
    The Mediator interface declares a method used by components to notify the
    mediator about various events. The Mediator may react to these events and
    pass the execution to other components.
    """

    @abstractmethod
    def notify(self, sender: BaseComponent, event: str) -> None:
        pass


class ConcreteMediator(Mediator):
    def __init__(self, component_1: Component1, component_2: Component2) -> None:
        self._component_1 = component_1
        self._component_1.mediator = self
        self._component_2 = component_2
        self._component_2.mediator = self

    @overrides
    def notify(self, sender: BaseComponent, event: str) -> None:
        if event == 'A':
            print('Mediator reacts on A and triggers following operations:')
            self._component_2.do_c()
        elif event == 'D':
            print('Mediator reacts on D and triggers following operations:')
            self._component_1.do_b()
            self._component_2.do_c()
